cubic-server/Player.cpp

```diff
@@ -575,19 +589,31 @@ void Player::_onSetPlayerPosition(const std::shared_ptr<protocol::SetPlayerPosit
 {
     LDEBUG("Got a Set Player Position (", pck->x, ", ", pck->feet_y, ", ", pck->z, ")");
     // TODO: Validate the position
+    if (pck->on_ground && _isFlying)
+        _isFlying = false;
+    if (!pck->on_ground && !_isJumping && !_isFlying && ((pck->feet_y + 64) - (_pos.y + 64) > 0)) {
+        _isJumping = true;
+        _isFlying = true;
+    }
     this->setPosition(pck->x, pck->feet_y, pck->z);
 }
 
 void Player::_onSetPlayerPositionAndRotation(const std::shared_ptr<protocol::SetPlayerPositionAndRotation> &pck)
 {
     LDEBUG("Got a Set Player Position And Rotation (", pck->x, ", ", pck->feet_y, ", ", pck->z, ")");
     // TODO: Validate the position
-    this->setPosition(pck->x, pck->feet_y, pck->z);
     float yaw_tmp = pck->yaw;
     while (yaw_tmp < 0)
         yaw_tmp += 360;
     while (yaw_tmp > 360)
         yaw_tmp -= 360;
+    if (pck->on_ground && _isFlying)
+        _isFlying = false;
+    if (!pck->on_ground && !_isJumping && !_isFlying && ((pck->feet_y + 64) - (_pos.y + 64) > 0)) {
```

Comment:
Same here

Reply:
You could create a function `_onMove` or something similar to avoid repeating code.